artichoke-backend/src/extn/core/time/trampoline.rs

```diff
@@ -472,10 +467,30 @@ pub fn subsec(interp: &mut Artichoke, time: Value) -> Result<Value, Error> {
 
 // Time format
 
-pub fn strftime(interp: &mut Artichoke, time: Value, format: Value) -> Result<Value, Error> {
-    let _ = interp;
-    let _ = time;
-    let _ = format;
-    // Requires a parser.
-    Err(NotImplementedError::new().into())
+pub fn strftime(interp: &mut Artichoke, mut time: Value, format: Value) -> Result<Value, Error> {
+    let time = unsafe { Time::unbox_from_value(&mut time, interp)? };
+    let format = interp.try_convert_mut(format)?;
+
+    let bytes: Vec<u8> = time.strftime(format).map_err(|e| {
+        // InvalidFormatString is the only true ArgumentError, where as the
+        // rest that can be thrown from strftime are runtime failures.
+        //
+        // ```console
+        // [3.1.2]> Time.now.strftime("%")
+        // (irb):1:in `strftime': invalid format: % (ArgumentError)
+        //      from (irb):1:in `<main>'
+        //	    from /home/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
+        //	    from /home/ben/.rbenv/versions/3.1.2/bin/irb:25:in `load'
+        //	    from /home/ben/.rbenv/versions/3.1.2/bin/irb:25:in `<main>'
+        // ```
+        if let strftime::Error::InvalidFormatString = e {
+            let mut message = br#"invalid format: "#.to_vec();
+            message.extend_from_slice(format);
+            Error::from(ArgumentError::from(message))
+        } else {
+            Error::from(RuntimeError::with_message("Unexpected failure"))
+        }
+    })?;
+
+    interp.try_convert_mut(&bytes[..])
```

Comment:
Couple of things:

When using the converters to string, passing the owned vec if possible saves an allocation.

I think we need to preserve the encoding of the format string in the output string which means we need to unbox the format as a spinoso string directly (after getting a value using the new `to_str` MRI converter?) and then allocate and box a spinoso string with the correct encoding.

Reply:
This should all be achievable by constructing a String with particular encoding with the Vec right?

Specifically on boxing the owned Vec, I noticed there wasnt a `try_convert_mut` for `Vec<u8>`, thus this slice conversion. I assume in future we might even want to reduce and/stop usage of that kind of conversion right?

Reply:
there is a `TryConvertMut` for vec:

https://github.com/artichoke/artichoke/blob/b261606b1de84f7ce4670e7ddda53f5738cde6cb/artichoke-backend/src/convert/bytes.rs#L14-L22

But yes long term I'd love to move in the direction of always going through spinoso-string. Here's an example of some bugs we've had due to these encoding issues:

- fac36b9f45b036ef2535673a95064afcc2776f73
- https://github.com/artichoke/artichoke/pull/2089

Reply:
I'm mostly certain that this is fixed now. I'll resolve this for now, but I'll put a question in a place which I'm not 100% sure about it's safety.